### PR TITLE
Add linear matrix HDF output

### PR DESCRIPTION
**Pull request by markdewing**
Add parameter "output_matrices_hdf" to output the linear method matrices in HDF format.
Also add information (shifts, etc) and outputs (eigenvalue, eigenvector) to reproduce the generalized eigensolver's output and check alternate implementations against QMCPACK's implementation.

The file is named "linear_matrices.h5".  Currently there is no support for using the root project name or the series number.

There are two uses for this output.
1. Explore alternate solvers and implementations outside of QMCPACK, especially for scaling with number of variational parameters.
2. A future PR may also write out the inputs to `fillOverlapAndHamiltonianMatrices` to another HDF file and these can be used to regression test that function.

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_
- New feature
- Testing changes (e.g. new unit/integration/performance tests)

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?
laptop and desktop

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- No. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

### CODE REVIEW

**Comment by markdewing** [comment]
[qmcpack_linear_solver_python.tar.gz](https://github.com/QMCPACK/qmcpack/files/6911735/qmcpack_linear_solver_python.tar.gz)

Attached Python script to read in the HDF file, run a solver, and compare with QMCPACK values.

**Comment by prckent** [comment]
Very minor request, please can you add an entry in the manual for this new parameter?

File: src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.h [hunk 1]
```diff
@@ -207,6 +207,9 @@ class QMCFixedSampleLinearOptimizeBatched : public QMCLinearOptimizeBatched,
   // Output Hamiltonian and overlap matrices
   bool do_output_matrices_;
 
+  // Output Hamiltonian and overlap matrices in HDF format
+  bool do_output_matrices_hdf_;
```
Comment (ye-luo) [inline src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.h:211]: is do_output_matrices_ a master control of do_output_matrices_hdf_?
If two independent switches for text and hdf5 output are needed, rename the old one to be more explicit.

File: src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp [hunk 2]
```diff
@@ -471,20 +472,23 @@ void QMCFixedSampleLinearOptimizeBatched::process(xmlNodePtr q)
...
   std::string OutputMatrices("no");
+  std::string OutputMatricesHDF("no");
   std::string FreezeParameters("no");
   OhmmsAttributeSet oAttrib;
   oAttrib.add(useGPU, "gpu");
   oAttrib.add(vmcMove, "move");
   oAttrib.add(ReportToH5, "hdf5");
 
   m_param.add(OutputMatrices, "output_matrices");
+  m_param.add(OutputMatricesHDF, "output_matrices_hdf");
```
Comment (ye-luo) [inline src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp:483]: Please modernize all the "add" as
https://github.com/QMCPACK/qmcpack/blob/9dfd387e343ed0ecab8583a7de4c26baba225794/src/io/OhmmsData/AttributeSet.h#L58
Use the third argument to restrict input options.

File: src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp [hunk 3]
```diff
@@ -471,20 +472,23 @@ void QMCFixedSampleLinearOptimizeBatched::process(xmlNodePtr q)
...
   m_param.add(FreezeParameters, "freeze_parameters");
 
   oAttrib.put(q);
   m_param.put(q);
 
-  do_output_matrices_ = (OutputMatrices != "no");
-  freeze_parameters_  = (FreezeParameters != "no");
+  do_output_matrices_     = (OutputMatrices != "no");
+  do_output_matrices_hdf_ = (OutputMatricesHDF != "no");
+  freeze_parameters_      = (FreezeParameters != "no");
```
Comment (ye-luo) [inline src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp:491]: Please clean up as `add()` has more restriction.

File: src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.h (hunk 1 above)
Comment (markdewing) [reply src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.h:211]: They are independent.  The original switch was named `do_output_matrices_csv_`

File: src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp (hunk 2 above)
Comment (markdewing) [reply src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp]: Fixed

File: src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp (hunk 3 above)
Comment (markdewing) [reply src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp]: Fixed

**Review by ye-luo — COMMENTED** [review]
LGTM. Need non-ANL approval

**Comment by prckent** [comment]
How difficult is it to get the root project name or the series number? I don't think that it is critical since this is more of a debug option, but it would be nice to have. We already write a few files by default that aren't labeled, but those are likely to be removed in a future version or renamed.

**Review by prckent — DISMISSED** [review]
If the naming is easy to change, please do so. Otherwise I think we should merge and keep moving.

**Comment by markdewing** [comment]
I can add the root name (project name + series number) - it's easier to get than I thought.

**Review by prckent — APPROVED** [review]
Thanks Mark!

**Comment by ye-luo** [comment]
Start testing in-house

**Pull request merged**